freezegun/api.py

```diff
@@ -37,6 +41,70 @@
     import copyreg
 
 
+# keep a cache of module attributes otherwise freezegun will need to analyze too many modules all the time
+# start with `None` as the sentinel value.
+# if `{}` (empty dict) was the sentinel value, there's a chance that `setup_modules_cache()` will be called many times
+_GLOBAL_MODULES_CACHE = None
+
+
+def _get_global_modules_cache():
+    global _GLOBAL_MODULES_CACHE
+    # the first call to this function sets up the global module cache. it's expected to be slower than consecutive calls
+    if _GLOBAL_MODULES_CACHE is None:
+        _GLOBAL_MODULES_CACHE = {}
+        _setup_modules_cache()
+    return _GLOBAL_MODULES_CACHE
+
+
+def _get_module_attributes(module):
+    result = []
+    try:
+        module_attributes = dir(module)
+    except TypeError:
+        return result
+    for attribute_name in module_attributes:
+        try:
+            attribute_value = getattr(module, attribute_name)
+        except (ImportError, AttributeError, TypeError):
+            # For certain libraries, this can result in ImportError(_winreg) or AttributeError (celery)
+            continue
+        else:
+            result.append((attribute_name, attribute_value))
+    return result
+
+
+def _setup_modules_cache():
+    for mod_name, module in list(sys.modules.items()):
+        # ignore modules from freezegun
+        if mod_name == __name__ or not mod_name or not module or not hasattr(module, "__name__"):
+            continue
+        _setup_module_cache(module)
+
+
+def _setup_module_cache(module):
+    global _GLOBAL_MODULES_CACHE
+    date_attrs = []
+    all_module_attributes = _get_module_attributes(module)
+    for attribute_name, attribute_value in all_module_attributes:
+        if id(attribute_value) in _real_time_object_ids:
+            date_attrs.append((attribute_name, attribute_value))
+    _GLOBAL_MODULES_CACHE[module.__name__] = (_get_module_attributes_hash(module), date_attrs)
+
+
+def _get_module_attributes_hash(module):
+    return '{0}-{1}'.format(id(module), hashlib.md5(','.join(dir(module)).encode('utf-8')).hexdigest())
```

Comment:
In line 63, you catch a `TypeError` when you call `dir(module)`, here you don't. If a `TypeError` is raised there, that doesn't seem to prevent this function from being called. Is that an issue?
Why don't you pass the already existing `all_module_attributes` to this function and extract the names from that?

Reply:
Thanks, fixed with https://github.com/spulec/freezegun/commit/689a7bbca2f1d39e5e35c83664f5e6fc73b07f40.